front/app/modules/commercial/granular_permissions/admin/components/UserFieldSelection/UserFieldSelection.tsx

```diff
@@ -0,0 +1,160 @@
+import React, { useState } from 'react';
+
+// intl
+import messages from '../../containers/Granular/messages';
+
+// components
+import {
+  Title,
+  Text,
+  Button,
+  Box,
+  colors,
+  Toggle,
+} from '@citizenlab/cl2-component-library';
+import { FieldSelectionModal } from './FieldSelectionModal';
+
+// api
+import useAddPermissionCustomField from 'api/permissions_custom_fields/useAddPermissionsCustomField';
+import usePermissionsCustomFields from 'api/permissions_custom_fields/usePermissionsCustomFields';
+import useDeletePermissionsCustomField from 'api/permissions_custom_fields/useDeletePermissionsCustomField';
+import useUpdatePermissionsCustomField from 'api/permissions_custom_fields/useUpdatePermissionsCustomField';
+
+// utils
+import { FormattedMessage } from 'utils/cl-intl';
+import FormattedMessageComponent from 'utils/cl-intl/FormattedMessage';
+
+// services
+import { IPermissionData } from 'services/actionPermissions';
+import useUserCustomFields from 'hooks/useUserCustomFields';
+import useLocale from 'hooks/useLocale';
+import { isNilOrError } from 'utils/helperUtils';
+import { IUserCustomFieldData } from 'services/userCustomFields';
+
+type UserFieldSelectionProps = {
+  permission: IPermissionData;
+  projectId?: string | null;
+  phaseId?: string | null;
+  initiativeContext?: boolean;
+};
+
+const UserFieldSelection = ({
+  permission,
+  projectId,
+  phaseId,
+  initiativeContext,
+}: UserFieldSelectionProps) => {
+  const globalRegistrationFields = useUserCustomFields();
+  const initialFields = usePermissionsCustomFields({
+    projectId,
+    phaseId,
+    initiativeContext,
+    action: permission.attributes.action,
+  });
+  const { mutate: addPermissionCustomField } = useAddPermissionCustomField();
+  const { mutate: updatePermissionCustomField } =
+    useUpdatePermissionsCustomField();
+  const { mutate: deletePermissionsCustomField } =
+    useDeletePermissionsCustomField();
+  const locale = useLocale();
+  const [showSelectionModal, setShowSelectionModal] = useState(false);
+  const initialFieldArray = initialFields?.data?.data;
+
+  const handleAddField = (field: IUserCustomFieldData) => {
```

Comment:
This operation is a bit slow on my local machine, maybe we could add a loading state?